# Spotify Data

## Index

* [Introduction](#introduction)
* [Data description](#description)
* [Questions](#questions)

## Introduction <a class="anchor" id="introduction"></a>

The dataset I will work in this notebook is a dataset of *Spotify* songs with several different attributes. I think this dataset could be useful to answer some questions about musical genres, popularity of songs and artists. The data was found on the *Kaggle* datasets repository and can be downloaded by pressing this [link](https://www.kaggle.com/edalrami/19000-spotify-songs#song_data.csv). Let first get started by loading the data and describing how it looks like.

## Data description <a class="anchor" id="description"></a>

As stated in the introduction, this dataset was taken from *Kaggle* and it's composed by different kinds of information for 19000 *Spotify* songs. We find that we have 2 diferent *CSV* files with different information about songs, so first of all, let's load the files into this notebook.

In [11]:
import pandas as pd

songs_df = pd.read_csv("./data/song_data.csv", index_col=0)
info_df = pd.read_csv("./data/song_info.csv", index_col=0)

Now let's look how these two dataframes look like so we can have an idea of the information inside them.

In [12]:
songs_df.head()

song_popularity  song_duration_ms  acousticness  \
song_name                                                                     
Boulevard of Broken Dreams               73            262333      0.005520   
In The End                               66            216933      0.010300   
Seven Nation Army                        76            231733      0.008170   
By The Way                               74            216933      0.026400   
How You Remind Me                        56            223826      0.000954   

                            danceability  energy  instrumentalness  key  \
song_name                                                                 
Boulevard of Broken Dreams         0.496   0.682          0.000029    8   
In The End                         0.542   0.853          0.000000    3   
Seven Nation Army                  0.737   0.463          0.447000    0   
By The Way                         0.451   0.970          0.003550    0   
How You Remind Me                  0.447   0.766          0.000000   10   

                            liveness  loudness  audio_mode  speechiness  \
song_name                                                                 
Boulevard of Broken Dreams    0.0589    -4.095           1       0.0294   
In The End                    0.1080    -6.407           0       0.0498   
Seven Nation Army             0.2550    -7.828           1       0.0792   
By The Way                    0.1020    -4.938           1       0.1070   
How You Remind Me             0.1130    -5.065           1       0.0313   

                              tempo  time_signature  audio_valence  
song_name                                                           
Boulevard of Broken Dreams  167.060               4          0.474  
In The End                  105.256               4          0.370  
Seven Nation Army           123.881               4          0.324  
By The Way                  122.444               4          0.198  
How You Remind Me           172.011               4          0.574

In [13]:
info_df.head()

artist_name  \
song_name                                           
Boulevard of Broken Dreams              Green Day   
In The End                            Linkin Park   
Seven Nation Army               The White Stripes   
By The Way                  Red Hot Chili Peppers   
How You Remind Me                      Nickelback   

                                                   album_names  \
song_name                                                        
Boulevard of Broken Dreams  Greatest Hits: God's Favorite Band   
In The End                                       Hybrid Theory   
Seven Nation Army                                     Elephant   
By The Way                         By The Way (Deluxe Version)   
How You Remind Me                               Silver Side Up   

                                    playlist  
song_name                                     
Boulevard of Broken Dreams  00s Rock Anthems  
In The End                  00s Rock Anthems  
Seven Nation Army           00s Rock Anthems  
By The Way                  00s Rock Anthems  
How You Remind Me           00s Rock Anthems

As we can see, in the dataset named *songs_df* we can find several attributes that we list in the next lines. I found the description of some of them in this [link](https://towardsdatascience.com/is-my-spotify-music-boring-an-analysis-involving-music-data-and-machine-learning-47550ae931de) and this [link](https://developer.spotify.com/documentation/web-api/reference/tracks/get-audio-features/):

- **Song name**: This is the song name, it serves as the index of the dataframe.

- **Song popularity**: A number indicating how popular a song is. The higer the number the most popular the song is.

- **Song duration ms***: A number with the duration of the song in miliseconds.

- **Acousticness**: This value describes how acoustic a song is. A score of 1.0 means the song is most likely to be an acoustic one.

- **Danceability**: Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.

- **Energy**: Represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy.

- **Instrumentalness**: This value represents the amount of vocals in the song. The closer it is to 1.0, the more instrumental the song is.

- **Key**: The estimated overall key of the track. Integers map to pitches using standard Pitch Class notation . E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.

- **Liveness**: This value describes the probability that the song was recorded with a live audience. According to the official documentation “a value above 0.8 provides strong likelihood that the track is live”.

- **Loudness**: The overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typical range between -60 and 0 db.

- **Audio mode**: Mode indicates the modality (major or minor) of a track, the type of scale from which its melodic content is derived. Major is represented by 1 and minor is 0.

- **Spechiness**: “Speechiness detects the presence of spoken words in a track”. If the speechiness of a song is above 0.66, it is probably made of spoken words, a score between 0.33 and 0.66 is a song that may contain both music and words, and a score below 0.33 means the song does not have any speech.

- **Tempo**: The overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.

- **Time signature**: An estimated overall time signature of a track. The time signature (meter) is a notational convention to specify how many beats are in each bar (or measure).

- **Audio valence**: A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).

On the other side, the *info_df* contains the following attributes:

- **Song name**: This is the song name, it serves as the index of the dataframe. We can link this dataframe with the other one through this field.

- **Artist name**: Artist whose the song belongs to.

- **Album name**: Name of the album containing the song.

- **Playlist**: Playlist on *Spotify* from where the song was retrieved.

Now that we have defined the different features on the data, let's try to find more information about the data.

In [15]:
songs_df.describe()

song_popularity  song_duration_ms  acousticness  danceability  \
count     18835.000000      1.883500e+04  18835.000000  18835.000000   
mean         52.991877      2.182116e+05      0.258539      0.633348   
std          21.905654      5.988754e+04      0.288719      0.156723   
min           0.000000      1.200000e+04      0.000001      0.000000   
25%          40.000000      1.843395e+05      0.024100      0.533000   
50%          56.000000      2.113060e+05      0.132000      0.645000   
75%          69.000000      2.428440e+05      0.424000      0.748000   
max         100.000000      1.799346e+06      0.996000      0.987000   

             energy  instrumentalness           key      liveness  \
count  18835.000000      18835.000000  18835.000000  18835.000000   
mean       0.644995          0.078008      5.289196      0.179650   
std        0.214101          0.221591      3.614595      0.143984   
min        0.001070          0.000000      0.000000      0.010900   
25%        0.510000          0.000000      2.000000      0.092900   
50%        0.674000          0.000011      5.000000      0.122000   
75%        0.815000          0.002570      8.000000      0.221000   
max        0.999000          0.997000     11.000000      0.986000   

           loudness    audio_mode   speechiness         tempo  time_signature  \
count  18835.000000  18835.000000  18835.000000  18835.000000    18835.000000   
mean      -7.447435      0.628139      0.102099    121.073154        3.959119   
std        3.827831      0.483314      0.104378     28.714456        0.298533   
min      -38.768000      0.000000      0.000000      0.000000        0.000000   
25%       -9.044000      0.000000      0.037800     98.368000        4.000000   
50%       -6.555000      1.000000      0.055500    120.013000        4.000000   
75%       -4.908000      1.000000      0.119000    139.931000        4.000000   
max        1.585000      1.000000      0.941000    242.318000        5.000000   

       audio_valence  
count   18835.000000  
mean        0.527967  
std         0.244632  
min         0.000000  
25%         0.335000  
50%         0.527000  
75%         0.725000  
max         0.984000

In the summary shown in the previous line we can see many interesting things. The most interesting thing for me in this case is that we have data for **18835** songs and all the data is complete! We don't have any missing value for any of the features, which is really nice but usually this doesn't happens. In case we find missing value we need to decide what to do with them. We have several options depending on how many missing values we can find. The different options are: 

- **Deleting columns**: If there are many missing values in a column, that column would not be useful so we can delete it.

- **Deleting rows**: If we just have some missing values we can choose between deleting the affected rows or imputing them.

- **Imputing values**: Imputing value means filling these missing values with something. We could use for example the mean value for the entire column or we could use some more complicated methods like try to predict these values. When it's possible I prefer to use this method since we don't miss any information on this way.

Also, in the previous summary we can see the range of values for each variable (i.e. popularity ranges from 0 to 100), mininum and maximum values or how spread they are.

Now let's try to see which kind of data contains is each of these features.

In [18]:
songs_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18835 entries, Boulevard of Broken Dreams to Up to Me
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   song_popularity   18835 non-null  int64  
 1   song_duration_ms  18835 non-null  int64  
 2   acousticness      18835 non-null  float64
 3   danceability      18835 non-null  float64
 4   energy            18835 non-null  float64
 5   instrumentalness  18835 non-null  float64
 6   key               18835 non-null  int64  
 7   liveness          18835 non-null  float64
 8   loudness          18835 non-null  float64
 9   audio_mode        18835 non-null  int64  
 10  speechiness       18835 non-null  float64
 11  tempo             18835 non-null  float64
 12  time_signature    18835 non-null  int64  
 13  audio_valence     18835 non-null  float64
dtypes: float64(9), int64(5)
memory usage: 2.2+ MB


All the data in our data frame is numeric data, some of them are *integers* and some of them are *float* but we don't have any categorical feature except the index. This data set is pretty clean and we won't need to do much pre-process on the data, but we need to take into account that as stated previously, this is not usual.

## Questions <a class="anchor" id="questions"></a>

Now let's think in some interesting questions that would be able to be answered with the available data.
The first of them of course is to predict the popularity of a song based on its features. But before this I would like to answer some other questions. 